sw/device/silicon_creator/lib/drivers/otbn.c

```diff
@@ -165,58 +191,56 @@ rom_error_t otbn_imem_sec_wipe(void) {
-  HARDENED_RETURN_IF_ERROR(
-      check_offset_len(offset_bytes, num_words, kOtbnIMemSizeBytes));
-  otbn_write(kBase + OTBN_IMEM_REG_OFFSET + offset_bytes, src, num_words);
-  return kErrorOk;
-}
-
 rom_error_t otbn_dmem_sec_wipe(void) {
   return otbn_cmd_run(kOtbnCmdSecWipeDmem, kErrorOtbnSecWipeDmemFailed);
 }
 
-rom_error_t otbn_dmem_write(uint32_t offset_bytes, const uint32_t *src,
-                            size_t num_words) {
-  HARDENED_RETURN_IF_ERROR(
-      check_offset_len(offset_bytes, num_words, kOtbnDMemSizeBytes));
-  otbn_write(kBase + OTBN_DMEM_REG_OFFSET + offset_bytes, src, num_words);
-  return kErrorOk;
+/**
+ * Checks if the OTBN application's IMEM and DMEM address parameters are valid.
+ *
+ * IMEM and DMEM ranges must not be "backwards" in memory, with the end address
+ * coming before the start address, and the IMEM range must additionally be
+ * non-empty. Finally, separate sections in DMEM must not overlap each other
+ * when converted to DMEM address space.
+ *
+ * @param app the OTBN application to check
+ * @return true if the addresses are valid, otherwise false.
+ */
+static bool check_app_address_ranges(const otbn_app_t *app) {
+  // IMEM must have a strictly positive range (cannot be backwards or empty)
+  if (app->imem_end <= app->imem_start) {
+    return false;
+  }
+  // DMEM data section must not be backwards
+  if (app->dmem_data_end < app->dmem_data_start) {
+    return false;
+  }
+  return true;
 }
 
-rom_error_t otbn_dmem_read(uint32_t offset_bytes, uint32_t *dest,
-                           size_t num_words) {
-  HARDENED_RETURN_IF_ERROR(
-      check_offset_len(offset_bytes, num_words, kOtbnDMemSizeBytes));
-
-  size_t i = 0;
-  for (; launder32(i) < num_words; ++i) {
-    dest[i] = abs_mmio_read32(kBase + OTBN_DMEM_REG_OFFSET + offset_bytes +
-                              i * sizeof(uint32_t));
+rom_error_t otbn_load_app(const otbn_app_t app) {
+  if (!check_app_address_ranges(&app)) {
+    return kErrorOtbnInvalidArgument;
   }
-  HARDENED_CHECK_EQ(i, num_words);
 
-  return kErrorOk;
-}
+  // If OTBN is busy, wait for it to be done.
+  HARDENED_RETURN_IF_ERROR(otbn_busy_wait_for_done());
 
-void otbn_zero_dmem(void) {
-  size_t i = 0;
-  for (; launder32(i) < kOtbnDMemSizeBytes; i += sizeof(uint32_t)) {
-    abs_mmio_write32(kBase + OTBN_DMEM_REG_OFFSET + i, 0u);
-  }
-  HARDENED_CHECK_EQ(i, kOtbnDMemSizeBytes);
-}
+  // Wipe memories.
+  HARDENED_RETURN_IF_ERROR(otbn_dmem_sec_wipe());
+  HARDENED_RETURN_IF_ERROR(otbn_imem_sec_wipe());
 
-void otbn_set_ctrl_software_errs_fatal(bool enable) {
-  // Only one bit in the CTRL register so no need to read current value.
-  uint32_t new_ctrl;
+  const size_t imem_num_words = app.imem_end - app.imem_start;
+  const size_t data_num_words = app.dmem_data_end - app.dmem_data_start;
 
-  if (enable) {
-    new_ctrl = 1;
-  } else {
-    new_ctrl = 0;
-  }
+  // IMEM always starts at 0.
+  otbn_addr_t imem_start_addr = 0;
```

Comment:
nit: do you think we should keep the parameter for `imem_start` or should we just hardcode it in the function?

Reply:
There are two different address spaces at play here; `imem_start_addr` and the comment here refer to OTBN's address space, while the `imem_start` parameter is in Ibex's address space. (Variables with the type `otbn_addr_t` always reference OTBN's address space, although that type is just `uint32_t` under the hood.)